src/http/server.cr

```diff
@@ -95,72 +100,187 @@ class HTTP::Server
-    @processor = RequestProcessor.new(handler)
+  # DEPRECATED: Create an instance and call `#bind` instead.
+  def self.new(host : String, port : Int32, handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc)
+    new(handlers, &handler).tap do |server|
+      server.bind(host, port)
+    end
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler), &handler : Context ->)
-    handler = HTTP::Server.build_middleware handlers, handler
-    @processor = RequestProcessor.new(handler)
+  # DEPRECATED: Create an instance and call `#bind` instead.
+  def self.new(host : String, port : Int32, handlers : Array(HTTP::Handler))
+    new(handlers).tap do |server|
+      server.bind(host, port)
+    end
   end
 
-  def initialize(@host : String, @port : Int32, handlers : Array(HTTP::Handler))
-    handler = HTTP::Server.build_middleware handlers
-    @processor = RequestProcessor.new(handler)
+  # DEPRECATED: Create an instance and call `#bind` instead.
+  def self.new(host : String, port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)
+    new(handler).tap do |server|
+      server.bind(host, port)
+    end
+  end
+
+  # Creates a new HTTP server with the given block as handler.
+  def self.new(&handler : HTTP::Handler::Proc) : self
+    new(handler)
+  end
+
+  # Creates a new HTTP server with a handler chain constructed from the *handlers*
+  # array and the given block.
+  def self.new(handlers : Array(HTTP::Handler), &handler : HTTP::Handler::Proc) : self
+    new(HTTP::Server.build_middleware(handlers, handler))
+  end
+
+  # Creates a new HTTP server with the *handlers* array as handler chain.
+  def self.new(handlers : Array(HTTP::Handler)) : self
+    new(HTTP::Server.build_middleware(handlers))
   end
 
-  def initialize(@host : String, @port : Int32, handler : HTTP::Handler | HTTP::Handler::Proc)
+  # Creates a new HTTP server with the given *handler*.
+  def initialize(handler : HTTP::Handler | HTTP::Handler::Proc)
     @processor = RequestProcessor.new(handler)
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to `true` to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(host : String, port : Int32, reuse_port : Bool = false) : TCPServer
+    bind TCPServer.new(host, port, reuse_port: reuse_port)
+  end
+
+  # Creates a `TCPServer` listenting on `127.0.0.1` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(port : Int32, reuse_port : Bool = false) : TCPServer
+    bind "127.0.0.1", port, reuse_port
```

Comment:
I was gonna say that binding to `127.0.0.1` makes the `#bind(port)` overload somewhat useless (invisible on network, doesn't bind ipv6 interfaces) but that's how it's currently done...

That's still weird, thought, since `TCPServer.new(port)` will bind on `::` (all ipv4 & ipv6 interfaces) for example.

Reply:
Yes, this should probably be improved. Although I'm not entirely sure if `::` would be a sane default.
I'd suggest to keep the current behavior for now and open an issue about this to fix it once this is merged.

Reply:
I prefer to keep everything local-addressed by default, and force people to choose possibly insecure defaults (0.0.0.0, ::) by conscious decision.

Reply:
Still, we should make sure to take the IPv6 loopback `::1` into account as well. It would probably nice to have a constant or class method in `Socket::IPAddress` instead of using magic value here.

Reply:
Maybe we could have a method to listen on `0.0.0.0` and `::`, like `#bind_all(port : Int32)`